### PR TITLE
Introduces `Bytes`, `Blob`, and `Clob` wrapper types

### LINKED ISSUE #500: Value/Content APIs Factoring to Capture Structurally Similar Content Types
### Problem Statement

Ion contains types that are structurally identical, but differ only in their _type_.  In our current factoring of `Value`, we end up creating variants that are basically the same content, but differ only in the type of `struct` used in the variant.  This leads to some code duplication and traits, that require generic code to be used if one wants to operate across these structurally identical types.  I believe we could "push down" the value variants as a way to factor the API to avoid this boilerplate.

This is a distillation of my thinking after an offline thread with @zslayton.

### Background

In the original `Element`/`Value` APIs the `Value` was the `enum` representing _type_ and its _contents_ of the value were modeled directly in the `enum`.  The following is the original factoring abbreviated for illustration purposes:

```rs
struct Bytes {
    bytes: Vec<u8>,
}

struct Sequence {
    children: Vec<Element>,
}

enum Value {
    Int(Int),
    Float(f64),
    Decimal(Decimal),
    Timestamp(Timestamp),
    String(String),
    Symbol(Symbol),
    Bool(bool),
    Blob(Bytes),
    Clob(Bytes),
    SExp(Sequence),
    List(Sequence),
    Struct(Struct),
}
```

The latest refactoring of these APIs have captured the type within the _content_ for the variants that are useful as intermediaries for the various APIs that build up these types and need to capture the intent of the user (e.g., I would like to build this sequence that is to become a `Value::List(...)`)  this need is captured for example in the [builder APIs and macros PR][1].  This factoring looks something like this (again, abbreviated/digested for illustration purposes):

```rs
struct Blob {
    bytes: Vec<u8>,
}

struct Clob {
    bytes: Vec<u8>,
}

// trait to allow us to generically operate across Blob/Clob
trait Lob { ... }

struct SExp {
    children: Vec<Element>,
}

struct List {
    children: Vec<Element>,
}

// trait to allow us to generically operate across SExp/List
// called IonSequence in the real APIs
trait Sequence { ... }

enum Value {
    Int(Int),
    Float(f64),
    Decimal(Decimal),
    Timestamp(Timestamp),
    String(String),
    Symbol(Symbol),
    Bool(bool),
    Blob(Blob),
    Clob(Clob),
    SExp(SExp),
    List(List),
    Struct(Struct),
}
```

The problem with the above is two-fold, we have to sort of copy/paste the implementation `Clob`/`Blob` and `SExp`/`List`.  These types have no relationship to each other (modulo a shared trait) and have some minor ergonomic implications to matching on say any `Blob`/`Clob` (i.e., work on any "LOB" value).

### Proposed Factoring

I think we can mitigate the boilerplate above while retaining the benefits of having _content_ that is strongly typed and "knows" what kind of variant it needs to be by pushing down these shared types in the `Value` enum:

```rs
enum LobType {
    Blob,
    Clob,
}
struct Lob {
    lob_type: LobType,
    bytes: Vec<u8>,
}

enum SequenceType {
    List,
    SExp,
}
struct Sequence {
    sequence_type: SequenceType,
    children: Vec<Element>,
}

enum Value {
    Int(Int),
    Float(f64),
    Decimal(Decimal),
    Timestamp(Timestamp),
    String(String),
    Symbol(Symbol),
    Bool(bool),

    // single variant for CLOB/BLOB
    Lob(Lob),
    // single variant for List/SExp
    Sequence(Sequence),
    
    Struct(Struct),
}
```

In the above, de-structuring becomes a bit more nested and the representation is marginally bigger in memory, but we no longer need the bridge traits and code can construct `Lob` and `Sequence` independently from `Value` which eliminates the boilerplate and the need for the bridge trait.  This factoring also makes it trivial to operate on the structurally similar types by destructing to the common variant (i.e., `Lob(...)` and `Sequence(...)`).

[1]: https://github.com/amazon-ion/ion-rust/pull/479#discussion_r1134239633

### PR DESCRIPTION
**Pull request by zslayton**
This PR introduces new opaque wrappers for `Vec<u8>`, which was used throughout the APIs to represent owned blobs and clobs.

`Bytes` captures the core functionality of an owned, immutable byte array, but does not include Ion type information.

`Blob` and `Clob` are both trivial tuple struct wrappers around `Bytes`, adding an Ion type to the data.

The `Value::Blob` and `Value::Clob` enum variants (which already provide an Ion type) now wrap an instance of `Bytes`.

This PR fixes #500.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/binary/non_blocking/raw_binary_reader.rs [hunk 1]
```diff
@@ -769,8 +770,8 @@ impl<A: AsRef<[u8]>> IonReader for RawBinaryBufferReader<A> {
         self.read_symbol_id().map(RawSymbolToken::SymbolId)
     }
 
-    fn read_blob(&mut self) -> IonResult<Vec<u8>> {
-        self.read_blob_bytes().map(Vec::from)
+    fn read_blob(&mut self) -> IonResult<Blob> {
+        self.read_blob_bytes().map(Vec::from).map(Blob::from)
```
Comment (zslayton) [inline src/binary/non_blocking/raw_binary_reader.rs:774]: 🗺️ The `IonReader` trait's blob- and clob- related methods had to be updated to return `Blob` and `Clob`.

Before:
```rust
fn read_blob(&mut self) -> IonResult<Vec<u8>>;
fn read_clob(&mut self) -> IonResult<Vec<u8>>;
```

After:
```rust
fn read_blob(&mut self) -> IonResult<Blob>;
fn read_clob(&mut self) -> IonResult<Clob>;
```

This change had to be made in each implementation of `IonReader`, of which there are several.

File: src/element/lob.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Clob(pub Bytes);
+
+impl Clob {
+    pub fn as_slice(&self) -> &[u8] {
+        self.as_ref()
+    }
+}
+
+impl AsRef<[u8]> for Blob {
```
Comment (zslayton) [inline src/element/lob.rs:44]: 🗺️ There are a lot of trait impls here that feel repetitive. I'm confident that we can represent them more concisely if needed (perhaps via a macro or a `Lob` trait with blanket implementations), but I want to leave that for a future PR.

File: src/element/mod.rs [hunk 3]
```diff
@@ -64,8 +69,8 @@ pub enum Value {
     String(Str),
     Symbol(Symbol),
     Bool(bool),
-    Blob(Vec<u8>),
-    Clob(Vec<u8>),
+    Blob(Bytes),
+    Clob(Bytes),
```
Comment (zslayton) [inline src/element/mod.rs:73]: 🗺️ Notice that we can still handle blob and clob values in a single `match`/`if let` arm:
```rust
if let Blob(b) | Clob(b) = value {
  // ...do stuff with `b`...
}
```

File: src/element/mod.rs [hunk 4]
```diff
@@ -176,7 +181,21 @@ impl From<&[u8]> for Value {
...
+        let bytes: Bytes = blob.into();
+        Value::Blob(bytes)
+    }
+}
+
+impl From<Clob> for Value {
+    fn from(clob: Clob) -> Self {
+        let bytes: Bytes = clob.into();
+        Value::Clob(bytes)
     }
```
Comment (zslayton) [inline src/element/mod.rs:199]: 🗺️ The standalone `Blob` and `Clob` types can be trivially widened to `Value`.

File: src/element/reader.rs [hunk 5]
```diff
@@ -147,8 +147,8 @@ impl<'a, R: IonReader<Item = StreamItem, Symbol = Symbol>> ElementLoader<'a, R>
                     Timestamp => Value::Timestamp(self.reader.read_timestamp()?),
                     Symbol => Value::Symbol(self.reader.read_symbol()?),
                     String => Value::String(self.reader.read_string()?),
-                    Clob => Value::Clob(self.reader.read_clob()?),
-                    Blob => Value::Blob(self.reader.read_blob()?),
+                    Clob => Value::Clob(self.reader.read_clob()?.into()),
+                    Blob => Value::Blob(self.reader.read_blob()?.into()),
```
Comment (zslayton) [inline src/element/reader.rs:151]: 🗺️ The `.into()` here is converting the lobs into their inner `Bytes` representations. From a memory layout perspective, this is a no-op as `Bytes` is the only field in `Blob`/`Clob`.

**Review by zslayton — COMMENTED** [review]
🗺️ PR tour

**Review by popematt — APPROVED** [review]

**Pull request merged**